File: Examples/NativeExample/NativeExample/StarRatingView.m
/*
 *  StarRatingView.m
 *
 * This file is a part of the Yandex Advertising Network.
 *
 * Version for iOS © 2016 YANDEX
 *
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://legal.yandex.com/partner_ch/
 */

#import "StarRatingView.h"

static CGFloat const kStarRatingMargin = 2.f;

@interface StarRatingView ()

@property (nonatomic, strong) UIImage *starImage;
@property (nonatomic, strong) NSNumber *rating;
@property (nonatomic, copy) NSArray *starViews;
@property (nonatomic, copy) NSArray *starConstraints;

@end

@implementation StarRatingView

- (instancetype)initWithFrame:(CGRect)frame starImage:(UIImage *)starImage
{
    self = [super initWithFrame:frame];
    if (self != nil) {
        _starImage = starImage;
    }
    return self;
}

- (void)setRating:(nullable NSNumber *)rating
{
    if ([_rating isEqualToNumber:rating] == NO) {
        _rating = rating;
        [self update];
    }
}

- (NSUInteger)numberOfStars
{
    return (NSUInteger)ceil([self.rating doubleValue]);
}

- (void)update
{
    for (UIView *starView in self.starViews) {
        [starView removeFromSuperview];
    }
    NSMutableArray *starViews = [NSMutableArray array];
    NSUInteger numberOfViews = [self numberOfStars];
    for (int i = 0; i < numberOfViews; ++i) {
        CGFloat widthPercent = 1.f;
        if (i == numberOfViews - 1) {
            widthPercent = self.rating.doubleValue - (CGFloat)i;
        }
        UIImageView *imageView = [self starImageViewWithWidthPercent:widthPercent];
        [self addSubview:imageView];
        [starViews addObject:imageView];
    }

    self.starViews = starViews;
    [self setNeedsUpdateConstraints];
}

- (UIImageView *)starImageViewWithWidthPercent:(CGFloat)percent
{
    UIImageView *imageView = [[UIImageView alloc] initWithImage:[self croppedStarImageForWidthPercent:percent]];
    imageView.translatesAutoresizingMaskIntoConstraints = NO;
    imageView.contentMode = UIViewContentModeLeft;
    return imageView;
}

- (UIImage *)croppedStarImageForWidthPercent:(CGFloat)widthPercent
{
    CGFloat width = self.starImage.size.width * widthPercent * self.starImage.scale;
    CGFloat height = self.starImage.size.height * self.starImage.scale;
    CGRect cropRegion = CGRectMake(0.f, 0.f, width, height);
    CGImageRef subImage = CGImageCreateWithImageInRect(self.starImage.CGImage, cropRegion);
    UIImage *croppedImage = [UIImage imageWithCGImage:subImage
                                                scale:self.starImage.scale
                                          orientation:self.starImage.imageOrientation];
    return croppedImage;
}

- (void)updateConstraints
{
    [super updateConstraints];
    [self removeConstraints:self.starConstraints];
    if ([self numberOfStars] != 0) {
        NSMutableArray *constraints = [NSMutableArray array];

        NSArray *boundaryViewsConstraints = [self boundaryStarViewsConstraints];
        NSArray *viewsInterConstraints = [self starViewsInterConstraints];
        [constraints addObjectsFromArray:boundaryViewsConstraints];
        [constraints addObjectsFromArray:viewsInterConstraints];
        self.starConstraints = constraints;
        [self addConstraints:constraints];
    }
}

- (NSArray *)boundaryStarViewsConstraints
{
    UIView *firstView = self.starViews.firstObject;
    UIView *lastView = self.starViews.lastObject;
    NSDictionary *views = NSDictionaryOfVariableBindings(firstView, lastView);
    NSArray *firstHorizontal = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[firstView]"
                                                                       options:0
                                                                       metrics:nil
                                                                         views:views];
    NSArray *lastHorizontal = [NSLayoutConstraint constraintsWithVisualFormat:@"H:[lastView]|"
                                                                      options:0
                                                                      metrics:nil
                                                                        views:views];
    return [firstHorizontal arrayByAddingObjectsFromArray:lastHorizontal];
}

- (NSArray *)starViewsInterConstraints
{
    NSMutableArray *constraints = [NSMutableArray array];
    UIView *previousView = nil;
    for (UIView *view in self.starViews) {
        NSDictionary *views = NSDictionaryOfVariableBindings(view);
        if (previousView != nil) {
            views = NSDictionaryOfVariableBindings(previousView, view);
            NSDictionary *metrics = @{ @"margin" : @(kStarRatingMargin) };
            NSArray *horizontal = [NSLayoutConstraint constraintsWithVisualFormat:@"H:[previousView]-margin-[view]"
                                                                          options:0
                                                                          metrics:metrics
                                                                            views:views];
            [constraints addObjectsFromArray:horizontal];
        }
        NSArray *vertical = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[view]|"
                                                                    options:0
                                                                    metrics:nil
                                                                      views:views];
        NSLayoutConstraint *widthToHeightConstraint = [NSLayoutConstraint constraintWithItem:view
                                                                                   attribute:NSLayoutAttributeWidth
                                                                                   relatedBy:NSLayoutRelationEqual
                                                                                      toItem:view
                                                                                   attribute:NSLayoutAttributeHeight
                                                                                  multiplier:1.f
                                                                                    constant:0.f];
        [constraints addObjectsFromArray:vertical];
        [constraints addObject:widthToHeightConstraint];
        previousView = view;
    }
    return [constraints copy];
}

- (CGSize)intrinsicContentSize
{
    CGSize size = CGSizeZero;
    NSUInteger numberOfStars = [self numberOfStars];
    if (numberOfStars != 0) {
        CGFloat height = self.starImage.size.height;
        CGFloat width = numberOfStars * self.starImage.size.width + (numberOfStars - 1) * kStarRatingMargin;
        size = CGSizeMake(width, height);
    }
    return size;
}

@end

File: Examples/YandexMobileAdsExample/YandexMobileAdsExample/Common/Native/StarRatingView.swift
/*
 * Version for iOS © 2015–2025 YANDEX
 *
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://yandex.com/legal/mobileads_sdk_agreement/
 */

import YandexMobileAds

class StarRatingView: UIView {
    private let starSpace: CGFloat = 5

    private var _rating: Int?
    private var starViews = [UIImageView]()

    private func updateStarViews() {
        for image in starViews {
            image.removeFromSuperview()
        }

        starViews.removeAll()

        guard let rating = _rating, rating > 0 else { return }
        showStar(atLeft: self, attribute: .trailing)
        for _ in 1 ..< rating {
            showStar(atLeft: starViews.last!, attribute: .leading)
        }
    }

    private func showStar(atLeft view: UIView, attribute: NSLayoutConstraint.Attribute) {
        let starImageView = UIImageView(image: UIImage(named: "star"))
        starImageView.translatesAutoresizingMaskIntoConstraints = false
        self.addSubview(starImageView)
        starImageView.addConstraint(NSLayoutConstraint(item: starImageView,
                                                       attribute: .height,
                                                       relatedBy: .equal,
                                                       toItem: nil,
                                                       attribute: .height,
                                                       multiplier: 1,
                                                       constant: starImageView.frame.size.height))
        starImageView.addConstraint(NSLayoutConstraint(item: starImageView,
                                                       attribute: .width,
                                                       relatedBy: .equal,
                                                       toItem: nil,
                                                       attribute: .width,
                                                       multiplier: 1,
                                                       constant: starImageView.frame.size.width))
        self.addConstraint(NSLayoutConstraint(item: starImageView,
                                              attribute: .trailing,
                                              relatedBy: .equal,
                                              toItem: view,
                                              attribute: attribute,
                                              multiplier: 1,
                                              constant: 0))
        self.addConstraint(NSLayoutConstraint(item: starImageView,
                                              attribute: .top,
                                              relatedBy: .equal,
                                              toItem: self,
                                              attribute: .top,
                                              multiplier: 1,
                                              constant: starSpace))
        starViews.append(starImageView)
    }
}

extension StarRatingView: Rating {
    func setRating(_ rating: NSNumber?) {
        _rating = rating as? Int
        updateStarViews()
    }

    func rating() -> NSNumber? {
        guard let rating = _rating else { return nil }
        return NSNumber(value: rating)
    }
}
